### PR TITLE
[ML] adding new PUT trained model vocabulary endpoint

### PR DESCRIPTION
**Pull request by benwtrent**
This commit removes the ability to set the vocabulary location in the model config. 
This opts instead for sane defaults to be set and used. Wrapping this up in an
API.

The index is now always the internally managed `.ml-inference-native` index
and the document ID is always `<model_id>_vocabulary`

This API only works for pytorch/nlp type models.

### CODE REVIEW

File: docs/reference/ml/df-analytics/apis/put-trained-model-vocabulary.asciidoc [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+[[put-trained-model-vocabulary]]
+= Create trained model vocabulary API
+[subs="attributes"]
+++++
+<titleabbrev>Create trained model vocabulary</titleabbrev>
+++++
+
+Creates a trained model vocabulary.
+This is only supported on NLP type models.
+experimental::[]
```
Comment (lcawl) [inline docs/reference/ml/df-analytics/apis/put-trained-model-vocabulary.asciidoc:12]: I think this needs to be separated by a new line to show up properly:
```suggestion

experimental::[]
```

File: rest-api-spec/src/main/resources/rest-api-spec/api/ml.put_trained_model_vocabulary.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      "paths":[
+        {
+          "path":"/_ml/trained_models/{model_id}/vocabulary",
+          "methods":[
+            "PUT"
+          ],
+          "parts":{
+            "model_id":{
+              "type":"string",
+              "description":"The ID of the trained model for this definition part"
```
Comment (davidkyle) [inline rest-api-spec/src/main/resources/rest-api-spec/api/ml.put_trained_model_vocabulary.json:23]:
```suggestion
              "description":"The ID of the trained model for this vocabulary"
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelVocabularyAction.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state, ActionListener<AcknowledgedResponse> listener) {
+
+        ActionListener<TrainedModelConfig> configActionListener = ActionListener.wrap(config -> {
+            InferenceConfig inferenceConfig = config.getInferenceConfig();
+            if ((inferenceConfig instanceof NlpConfig) == false) {
+                listener.onFailure(
+                    new ElasticsearchStatusException(
+                        "cannot put vocabulary for model [{}] as it is not an nlp model",
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelVocabularyAction.java:74]: nlp or NLP?

It's NLP in the docs but there is not a precedent in code

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/FillMaskConfig.java [hunk 4]
```diff
@@ -38,7 +40,19 @@ public static FillMaskConfig fromXContentLenient(XContentParser parser) {
     private static ConstructingObjectParser<FillMaskConfig, Void> createParser(boolean ignoreUnknownFields) {
         ConstructingObjectParser<FillMaskConfig, Void> parser = new ConstructingObjectParser<>(NAME, ignoreUnknownFields,
             a -> new FillMaskConfig((VocabularyConfig) a[0], (Tokenization) a[1]));
-        parser.declareObject(ConstructingObjectParser.constructorArg(), VocabularyConfig.createParser(ignoreUnknownFields), VOCABULARY);
+        parser.declareObject(
+            ConstructingObjectParser.optionalConstructorArg(),
+            (p, c) -> {
+                if (ignoreUnknownFields == false) {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/FillMaskConfig.java:46]: ++ the caller can't set this but the code creates the config so it is visible

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**